pkg/app/piped/apistore/eventstore/store.go

```diff
@@ -177,3 +172,37 @@ func (s *store) GetLatest(ctx context.Context, name string, labels map[string]st
 	s.latestEvents[key] = resp.Event
 	return resp.Event, true
 }
+
+func (s *store) UpdateStatuses(ctx context.Context, latestEvents []model.Event) error {
+	es := make([]*pipedservice.ReportEventStatusesRequest_Event, 0, len(latestEvents))
+	for _, e := range latestEvents {
+		es = append(es, &pipedservice.ReportEventStatusesRequest_Event{
+			Id:                e.Id,
+			Status:            e.Status,
+			StatusDescription: e.StatusDescription,
+		})
+	}
+	if _, err := s.apiClient.ReportEventStatuses(ctx, &pipedservice.ReportEventStatusesRequest{Events: es}); err != nil {
+		return fmt.Errorf("failed to report event statuses: %w", err)
+	}
+
+	// Update cached events.
+	for _, e := range latestEvents {
+		s.mu.Lock()
+		cached, ok := s.latestEvents[e.EventKey]
+		if !ok {
+			s.latestEvents[e.EventKey] = &e
+			s.mu.Unlock()
+			continue
+		}
+		if cached.Id != e.Id {
+			// There is already an event newer than the given one.
+			s.mu.Unlock()
+			continue
+		}
+		s.latestEvents[e.EventKey].Status = e.Status
+		s.latestEvents[e.EventKey].StatusDescription = e.StatusDescription
+		s.mu.Unlock()
```

Comment:
Or, how about overriding the cache by events fetched from control-plane again?
```suggestion
		resp, err := s.apiClient.GetLatestEvent(ctx, &pipedservice.GetLatestEventRequest{
			Name:   e.Name,
			Labels: e.Labels,
		})
		if err != nil {
			return err
		}
		s.mu.Lock()
		cached, ok := s.latestEvents[e.EventKey]
		if !ok {
			s.latestEvents[e.EventKey] = resp.Event
			s.mu.Unlock()
			continue
		}
		if cached.Id != e.Id {
			// There is already an event newer than the given one.
			s.mu.Unlock()
			continue
		}
		s.latestEvents[e.EventKey].Status = resp.Event.Status
		s.latestEvents[e.EventKey].StatusDescription = resp.Event.StatusDescription
		s.mu.Unlock()
```

Reply:
@nghialv This solves this concern https://github.com/pipe-cd/pipecd/pull/3108#issuecomment-1021933373, right?

Reply:
I see. That is a nice idea but it will add some extra calls.
A better way is to use the returned events from `ReportEventStatuses` RPC to update the cache since they were updated by the server.

```
if _, err := s.apiClient.ReportEventStatuses(ctx, &pipedservice.ReportEventStatusesRequest{Events: es}); err != nil ```

A nice point of those ways is that we don't have to rewrite the logic of updating the event model what has already been done by the server.

But I think both of them still not be able to solve the root problem.
We will find a good solution for all local stores not just `event` store in the future.

Anyway, let's go with the current implementation for now. 🙌

Reply:
Alright!